File: src/main/java/net/maple/handlers/user/UserAbilityMassUpRequestHandler.java
package net.maple.handlers.user;

import client.Character;
import client.Client;
import client.player.StatType;
import net.maple.handlers.PacketHandler;
import util.packet.PacketReader;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class UserAbilityMassUpRequestHandler extends PacketHandler {

    @Override
    public void handlePacket(PacketReader reader, Client c) {
        Character chr = c.getCharacter();
        reader.readInteger(); // ?

        int count = reader.readInteger();
        Map<Integer, Integer> inc = new HashMap<>();
        int total = 0;

        for (int i = 0; i < count; i++) {
            int type = reader.readInteger();
            int amount = reader.readInteger();
            inc.put(type, amount);
            total += amount;
        }

        if (chr.getAp() < total) {
            c.close(this, "More stats than AP available");
            return;
        }

        List<StatType> statTypes = new ArrayList<>();
        statTypes.add(StatType.AP);
        chr.setAp(chr.getAp() - total);

        inc.forEach((type, amount) -> {
            if (type == StatType.STR.getStat()) {
                chr.setStrength(chr.getStrength() + amount);
                statTypes.add(StatType.STR);
            } else if (type == StatType.DEX.getStat()) {
                chr.setDexterity(chr.getDexterity() + amount);
                statTypes.add(StatType.DEX);
            } else if (type == StatType.INT.getStat()) {
                chr.setIntelligence(chr.getIntelligence() + amount);
                statTypes.add(StatType.INT);
            } else if (type == StatType.LUK.getStat()) {
                chr.setLuck(chr.getLuck() + amount);
                statTypes.add(StatType.LUK);
            }
        });

        chr.updateStats(statTypes, true);
    }
}

File: src/main/java/net/maple/handlers/user/UserAbilityMassUpRequestHandler.kt
package net.maple.handlers.user

import client.Client
import client.player.StatType
import net.maple.handlers.PacketHandler
import util.packet.PacketReader
import java.util.*
import kotlin.collections.HashMap
import kotlin.collections.set

class UserAbilityMassUpRequestHandler : PacketHandler {

    override fun handlePacket(reader: PacketReader, c: Client) {
        val chr = c.character

        reader.readInteger() // timestamp
        val count = reader.readInteger()

        val inc: MutableMap<Int, Int> = HashMap()
        var total = 0
        for (i in 0 until count) {
            val type = reader.readInteger()
            val amount = reader.readInteger()
            inc[type] = amount
            total += amount
        }

        if (chr.ap < total) {
            c.close(this, "More stats than AP available")
            return
        }

        chr.ap -= total
        inc.forEach {
            when (it.key) {
                StatType.STR.stat -> chr.strength += it.value
                StatType.DEX.stat -> chr.dexterity += it.value
                StatType.INT.stat -> chr.intelligence += it.value
                StatType.LUK.stat -> chr.luck += it.value
            }
        }
    }
}